### PR TITLE
Fields for unique constraints

### LINKED ISSUE #8345: Add possibility to use fields instead of column for unique constraint
Currently when using a unique constraint like:

```xml
<unique-constraints>
    <unique-constraint columns="locale,product_key" />
</unique-constraints>
```

you are forced to define the `column` name for the fields e.g.:

```xml
<field name="productKey" column="product_key" type="string" nullable="true" unique="true"/>
```

if you don't do this and somebody configure another `naming_strategy` for your entity it will not longer work as the column  does then not exist.

It would be a great feature if you could use unique-constraints based on field name and doctrine orm will read the column name from the configured field e.g.:

```xml
<unique-constraints>
    <unique-constraint fields="locale,productKey" />
</unique-constraints>

<!-- no column is needed here now -->
<field name="productKey" type="string" nullable="true" unique="true"/>
```

What do you think about this?

### PR DESCRIPTION
**Pull request by beberlei**
This replaces #8523 - since its not using a PR branch on the remote.

It fixes #8345 to implement support for field based configuration on `Index` and `UniqueConstraint` instead of column based.

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/MappingException.php [hunk 1]
```diff
@@ -933,4 +933,32 @@ public static function illegalOverrideOfInheritedProperty($className, $propertyN
             )
         );
     }
+
+    /**
+     * @return self
+     */
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/MappingException.php:939]: Let's use a native return type declaration here?

File: lib/Doctrine/ORM/Tools/SchemaTool.php [hunk 2]
```diff
@@ -137,6 +138,50 @@ private function processingNotRequired(
...
 
+    /**
+     * Resolves fields in index mapping to column names
+     *
+     * @param ClassMetadata $class
+     * @param mixed[]       $indexData index or unique constraint data
+     *
+     * @return string[] Column names from combined fields and columns mappings
+     */
+    private function getIndexColumns($class, array $indexData)
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/SchemaTool.php]:
```suggestion
    private function getIndexColumns(ClassMetadata $class, array $indexData): array
```

File: lib/Doctrine/ORM/Tools/SchemaTool.php [hunk 3]
```diff
@@ -309,13 +354,13 @@ static function (ClassMetadata $class) use ($idMapping): bool {
                         $indexData['flags'] = [];
                     }
 
-                    $table->addIndex($indexData['columns'], is_numeric($indexName) ? null : $indexName, (array) $indexData['flags'], $indexData['options'] ?? []);
+                    $table->addIndex($this->getIndexColumns($class, $indexData), is_numeric($indexName) ? null : $indexName, (array) $indexData['flags'], $indexData['options'] ?? []);
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/SchemaTool.php]:
```suggestion
                    $table->addIndex(
                                        $this->getIndexColumns($class, $indexData),
                                        is_numeric($indexName) ? null : $indexName,
                                        (array) $indexData['flags'],
                                        $indexData['options'] ?? []
                    );
```

File: lib/Doctrine/ORM/Mapping/MappingException.php (hunk 1 above)
Comment (beberlei) [reply lib/Doctrine/ORM/Mapping/MappingException.php:939]: we can do that in one big step when we introduce typehints more consistently, this is how all other methods in `MappingException`, i prefer the consistency for now. We can rector everything at some point

**Review by greg0ire — APPROVED** [review]

**Pull request merged**